mls-rs/src/tree_kem/mod.rs

```diff
@@ -344,6 +351,35 @@ impl TreeKemPublic {
         CP: CipherSuiteProvider,
     {
         // Apply removes (they commute with updates because they don't touch the same leaves)
+        // Self-removes take precedence
+        #[cfg(all(feature = "custom_proposal", feature = "self_remove_proposal"))]
+        for i in (0..proposal_bundle.self_removes.len()).rev() {
+            let index = match proposal_bundle.self_removes[i].sender {
+                crate::group::Sender::Member(idx) => LeafIndex(idx),
+                _ => continue,
+            };
+            let res = self.nodes.blank_leaf_node(index);
+
+            if res.is_ok() {
+                // This shouldn't fail if `blank_leaf_node` succedded.
+                self.nodes.blank_direct_path(index)?;
+            }
+
+            #[cfg(feature = "tree_index")]
+            if let Ok(old_leaf) = &res {
+                // If this fails, it's not because the proposal is bad.
+                let identity =
+                    identity(&old_leaf.signing_identity, id_provider, extensions).await?;
+
+                self.index.remove(old_leaf, &identity);
+            }
+
+            if proposal_bundle.self_removes[i].is_by_value() || !filter {
+                res?;
+            } else if res.is_err() {
+                proposal_bundle.remove::<SelfRemoveProposal>(i);
+            }
```

Comment:
This code is almost the same as below - can the duplication be avoided somehow (function / chain removes and self-removes)?

Reply:
Good point! Done.